### PR TITLE
Fea/nhwc

### PR DESCRIPTION
**Pull request by strint**
NHWC for resnet fp16 perf
### 要处理的支持NHWC的Op列表
- [x] nn.CropMirrorNormalize
- [x] nn.Conv2d
- [x] nn.FusedBatchNorm2d
- [x] nn.BatchNorm2d
- [x] n.LegacyMaxPool2d
- [x] nn.LegacyAvgPool2d

nn.init.kaiming_normal_先保留其data_format字段，也接受环境变量的控制。

### CODE REVIEW

File: python/oneflow/nn/modules/conv.py [hunk 1]
```diff
@@ -390,6 +393,7 @@ def __init__(
         groups: int = 1,
         bias: bool = True,
         padding_mode: str = "zeros",
+        data_format: str = "NCHW",
```
Comment (chengtbf) [inline python/oneflow/nn/modules/conv.py:396]: 如果给 Conv 增加 data_format 的接口，是不是就不跟 torch 对齐了？这个是什么时候讨论决定的呢？

Comment (strint) [reply python/oneflow/nn/modules/conv.py:396]: 这个只是为了对齐resnet 50 fp16的性能写的

File: python/oneflow/nn/modules/pooling.py [hunk 2]
```diff
@@ -121,6 +121,218 @@ def extra_repr(self) -> str:
...
+        if i in range(dhw_offset, dhw_offset + ndims):
+            pads_list.append(pad)
+        else:
+            assert pad == [0, 0]
+    return pads_list
+
+def calc_pool_padding(padding, dhw_offset, ndims):
+    if isinstance(padding, str):
+        padding = "SAME_LOWER" if padding.upper() == "SAME" else padding
+        assert padding.upper() in ["VALID", "SAME_LOWER", "SAME_UPPER"]
```
Comment (strint) [inline python/oneflow/nn/modules/pooling.py:150]: "VALID" 怎么配置，和直接传padding的关系是什么

File: python/oneflow/nn/modules/conv.py (hunk 1 above)
Comment (chengtbf) [reply python/oneflow/nn/modules/conv.py:396]: 这个分支应该还不是最终方案的分支吧

Comment (strint) [reply python/oneflow/nn/modules/conv.py:396]: 不是，接口还没按设计改，我今天改下接口。

**Comment by strint** [comment]
需要合并到新Module中，有什么不能合并的原因么
- nn.LegacyMaxPool2d
- nn.LegacyAvgPool2d

**Review by Flowingsun007 — APPROVED** [review]

File: python/oneflow/nn/modules/conv.py [hunk 3]
```diff
@@ -364,13 +367,30 @@ def __init__(
         self.padding = _pair(padding)
         self.dilation = _pair(dilation)
         self.groups = groups
+
+        if os.getenv("ONEFLOW_ENABLE_NHWC") == "1":
+            self.data_format = "NHWC"
+        else:
+            self.data_format = "NCHW"
+        if self.data_format == "NCHW":
+            self.channel_first = True
```
Comment (chengtbf) [inline python/oneflow/nn/modules/conv.py:376]: 这里为什么有类似功能的三个字段： data_format，channels_first 和 channel_pos 。是不是只要有最后一个就行了？（channel_pos）

File: python/oneflow/nn/modules/dataset.py [hunk 4]
```diff
@@ -202,7 +203,6 @@ class CropMirrorNormalize(Module):
     def __init__(
         self,
         color_space: str = "BGR",
-        output_layout: str = "NCHW",
```
Comment (chengtbf) [inline python/oneflow/nn/modules/dataset.py:205]: 这里不能直接删除吧。。。这属于对前向接口不兼容，如果有用户之前写了这个会不会挂掉？ 可以弄一个 deprecated 接口？ 就是提示用户如果传了这个值，就会弹 warning？ 尤其是如果这里传了 NHWC 参数，需要提示用户，这个参数不生效，且需要使用环境变量 
```python
os.environ[“ONEFLOW_ENABLE_NHWC”) = 1
```

File: python/oneflow/nn/modules/linear.py [hunk 5]
```diff
@@ -109,7 +109,10 @@ def reset_parameters(self) -> None:
     def forward(self, x):
         res = flow._C.matmul(x, self.weight, transpose_a=False, transpose_b=True)
         if self.bias is not None:
-            res += self.bias
+            if flow._oneflow_internal.lazy_mode.is_enabled:
+                res = flow._C.bias_add(res, self.bias)
```
Comment (chengtbf) [inline python/oneflow/nn/modules/linear.py:113]: bias add 和 broadcast add 性能不同？

File: python/oneflow/nn/modules/pooling.py (hunk 2 above)
Comment (chengtbf) [reply python/oneflow/nn/modules/pooling.py:154]: VALID 就是不 padding 吧

**Review by chengtbf — APPROVED** [review]

**Comment by strint** [comment]
@ouyangyu 已经实验，本分支性能测试和之前的最好结果一致。

File: python/oneflow/nn/modules/conv.py (hunk 3 above)
Comment (strint) [reply python/oneflow/nn/modules/conv.py:376]: 是的，我简化下。之前python 和 c++ api 各暴露了一个概念，现在可以不用了。

File: python/oneflow/nn/modules/dataset.py (hunk 4 above)
Comment (strint) [reply python/oneflow/nn/modules/dataset.py:205]: 好的。

File: python/oneflow/nn/modules/linear.py (hunk 5 above)
Comment (strint) [reply python/oneflow/nn/modules/linear.py:113]: 在resnet下，没有性能差异，是为了对齐写的。我先去掉吧。

File: python/oneflow/nn/modules/pooling.py (hunk 2 above)
Comment (strint) [reply python/oneflow/nn/modules/pooling.py:154]: 是的

File: python/oneflow/nn/modules/pooling.py [hunk 6]
```diff
@@ -465,27 +544,70 @@ def __init__(
...
+                kernel_size=self.kernel_size,
+                stride=self.stride,
+                padding=self.padding,
+                ceil_mode=self.ceil_mode,
+                count_include_pad=self.count_include_pad,
+                divisor_override=self.divisor_override,
+                data_format=self.channel_pos,
+            )
+        else:
+            return flow._C.avg_pool_2d(
```
Comment (hjchen2) [inline python/oneflow/nn/modules/pooling.py:601]: 这两个avg_pool_2d和avg_pool2d的命名顺便改一下，用data_format来区分一下？

Comment (strint) [reply python/oneflow/nn/modules/pooling.py:601]: 好的
